### ISSUE TITLE
@vcd/sdk exception, when using with vCD 10

### ISSUE THREAD

**Issue by jhiemer** [vmware-samples/vcd-ext-samples#10]
Hi,
when we work with vCD 10 and use the @vcd/sdk I get the following error during build time:

```bash
ERROR in node_modules/@vcd/sdk/common/container-hooks.d.ts(4,10): error TS2305: Module '"/Users/jhiemer/Development/osb-framework/vcd/angular6/vcd-marketplace/node_modules/@angular/core/core"' has no exported member 'OpaqueToken'.
```

Reason is OpaqueToken is replace by InjectionToken with Angular 6 and vCD 10 is based on Angular `ng-version="6.1.10"`

Will there be an update soon?

**Comment by jmoroski** [comment]
I've deployed several plugins to vCD 10 without issue that use the SDK. Are you getting this error when deployed to a vCD 10 environment, or when using the SDK in the context of a standalone application? I ask because this is one of the shims that we added in our bootstrapping layer for plugin compatibility:
`const angularCoreCompat = {
    ...angularCore,
    OpaqueToken: angularCore.InjectionToken
};`

Basically we re-export InjectionToken as OpaqueToken so that a plugin written for 9.1-9.7 will not just stop working in 10.

**Comment by jhiemer** [comment]
Hi Jeff,
see here:

```container-hooks.d.ts
/**
 * This is the currently supported - albeit very minimal - public SDK.
 */
import { OpaqueToken } from '@angular/core';
import { Observable } from 'rxjs';
/**
 * Wire in as a string.  Gives the root URL for API access (for example, the load balancer URL
 * or the single-cell URL).
 */
export declare const API_ROOT_URL: OpaqueToken;
/**
 * Wire in as a string.  Gives the root URL for the legacy Flex application.
 */
export declare const FLEX_APP_URL: OpaqueToken;
/**
 * Wire in as a string.  Gives the current scope of the VCD-UI.  As of current, this will be
 * either 'tenant' for the tenant portal, or 'service-provider' for the service-provider portal.
 */
export declare const SESSION_SCOPE: OpaqueToken;
/**
 * Wire in as a string.  Gives the unique name (not the display name) of the current tenant
 * organization that the VCD-UI is being used for.
 */
export declare const SESSION_ORGANIZATION: OpaqueToken;
/**
 * Wire in as a string.  Gives the UUID identifier of the current tenant
 * organization that the VCD-UI is being used for.
 */
export declare const SESSION_ORG_ID: OpaqueToken;
/**
 * Wire in as a string.  Gives the full root path for module assets (e.g. images, scripts, text files)
 */
export declare const EXTENSION_ASSET_URL: OpaqueToken;
/**
 * Wire in as a string.  Gives the Angular 2 route that the module is registered under.
 */
export declare const EXTENSION_ROUTE: OpaqueToken;
/**
 * Wire in as a boolean.  True if running under the SDK, false if running in production.
 */
export declare const SDK_MODE: OpaqueToken;
/**
 * Payload for registering a navigation menu entry.
 */
export interface ExtensionNavRegistration {
    /**
     * The router path to navigate to when selected.
     */
    path: string;
    /**
     * i18 key for name of extension, to be used as text for links
     */
    nameCode: string;
    /**
     * i18n key for description of extension, to be used on hover for links
     */
    descriptionCode: string;
}
/**
 * Dispatch this to the redux store to register a navigation menu.
 */
export interface ExtensionNavRegistrationAction {
    payload: ExtensionNavRegistration;
    readonly type: string;
}
export interface ExtensionNavRegistrationActionCtor {
    new (payload: ExtensionNavRegistration): ExtensionNavRegistrationAction;
}
export declare const ExtensionNavRegistrationAction: ExtensionNavRegistrationActionCtor;
/**
 * Inject this to access the authentication token.
 */
export interface AuthTokenHolderService {
    /**
     * The authentication token.
     */
    token: string;
}
export declare const AuthTokenHolderService: {
    token: string;
};
/**
 * This represents a menu item that can be clicked to perform an action on an entity (e.g. VM).
 */
export interface EntityActionExtensionMenuItem {
    /**
     * This is a unique URN identifying the action. This is so the component can be notified which action was requested.
     */
    urn: string;
    /**
     * This is the label for the menu item. It is not automatically translated.
     */
    text: string;
    /**
     * This is the enabled state for the menu item.
     */
    enabled: boolean;
    /**
     * This is the busy state for the menu item.
     */
    busy: boolean;
}
/**
 * This represents a sub-menu that contains other sub-menus or menu actions.
 */
export interface EntityActionExtensionSubmenu {
    /**
     * This is the label for the sub-menu. It is not automatically translated.
     */
    text: string;
    /**
     * This is an arbitrary, ordered list of EntityActionMenuItem.
     */
    children: EntityActionExtensionMenuItem[];
}
/**
 * This represents menu information for an entity action.
 */
export declare type EntityActionExtensionMenuEntry = EntityActionExtensionMenuItem | EntityActionExtensionSubmenu;
/**
 * Every component referenced by an entity action extension point must inherit from this.
 */
export declare abstract class _EntityActionExtensionComponent {
    /**
     * This returns is an Observable that defines the menu entry for this component. This is actively bound -
     * the component is able to change it in order to have the menu change (for example, to change busy
     * or enabled states).
     */
    abstract getMenuEntry(entityUrn: string): Observable<EntityActionExtensionMenuEntry>;
    /**
     * This field is a method that is called if a menu item is clicked while enabled. It will be called with the menu
     * item's URN and the URN of the entity that the action is being called for (e.g. the URN of the VM being edited).
     * It must return a single-shot (e.g. of, fromPromise or .first) Observable on completion. This Observable returns
     * a single value, refreshRequested. If this is true, the entity being edited will be immediately refreshed.
     * @param menuItemUrn the URN of the clicked menu item
     * @param entityUrnThe URN of the entity that the action is being called for
     */
    abstract performAction(menuItemUrn: string, entityUrn: string): Observable<{
        refreshRequested: boolean;
    }>;
}
export declare const EntityActionExtensionComponent: typeof _EntityActionExtensionComponent;
```

Led to `ERROR in node_modules/@vcd/sdk/common/container-hooks.d.ts(4,10): error TS2305: Module '"/Users/jhiemer/Development/osb-framework/vcd/angular6/vcd-marketplace/node_modules/@angular/core/core"' has no exported member 'OpaqueToken'.`

which I had to change to:

```
/**
 * This is the currently supported - albeit very minimal - public SDK.
 */
import { InjectionToken } from '@angular/core';
import { Observable } from 'rxjs';
/**
 * Wire in as a string.  Gives the root URL for API access (for example, the load balancer URL
 * or the single-cell URL).
 */
export declare const API_ROOT_URL: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the root URL for the legacy Flex application.
 */
export declare const FLEX_APP_URL: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the current scope of the VCD-UI.  As of current, this will be
 * either 'tenant' for the tenant portal, or 'service-provider' for the service-provider portal.
 */
export declare const SESSION_SCOPE: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the unique name (not the display name) of the current tenant
 * organization that the VCD-UI is being used for.
 */
export declare const SESSION_ORGANIZATION: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the UUID identifier of the current tenant
 * organization that the VCD-UI is being used for.
 */
export declare const SESSION_ORG_ID: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the full root path for module assets (e.g. images, scripts, text files)
 */
export declare const EXTENSION_ASSET_URL: InjectionToken<String>;
/**
 * Wire in as a string.  Gives the Angular 2 route that the module is registered under.
 */
export declare const EXTENSION_ROUTE: InjectionToken<String>;
/**
 * Wire in as a boolean.  True if running under the SDK, false if running in production.
 */
export declare const SDK_MODE: InjectionToken<String>;
/**
 * Payload for registering a navigation menu entry.
 */
export interface ExtensionNavRegistration {
    /**
     * The router path to navigate to when selected.
     */
    path: string;
    /**
     * i18 key for name of extension, to be used as text for links
     */
    nameCode: string;
    /**
     * i18n key for description of extension, to be used on hover for links
     */
    descriptionCode: string;
}
/**
 * Dispatch this to the redux store to register a navigation menu.
 */
export interface ExtensionNavRegistrationAction {
    payload: ExtensionNavRegistration;
    readonly type: string;
}
export interface ExtensionNavRegistrationActionCtor {
    new (payload: ExtensionNavRegistration): ExtensionNavRegistrationAction;
}
export declare const ExtensionNavRegistrationAction: ExtensionNavRegistrationActionCtor;
/**
 * Inject this to access the authentication token.
 */
export interface AuthTokenHolderService {
    /**
     * The authentication token.
     */
    token: string;
}
export declare const AuthTokenHolderService: {
    token: string;
};
/**
 * This represents a menu item that can be clicked to perform an action on an entity (e.g. VM).
 */
export interface EntityActionExtensionMenuItem {
    /**
     * This is a unique URN identifying the action. This is so the component can be notified which action was requested.
     */
    urn: string;
    /**
     * This is the label for the menu item. It is not automatically translated.
     */
    text: string;
    /**
     * This is the enabled state for the menu item.
     */
    enabled: boolean;
    /**
     * This is the busy state for the menu item.
     */
    busy: boolean;
}
/**
 * This represents a sub-menu that contains other sub-menus or menu actions.
 */
export interface EntityActionExtensionSubmenu {
    /**
     * This is the label for the sub-menu. It is not automatically translated.
     */
    text: string;
    /**
     * This is an arbitrary, ordered list of EntityActionMenuItem.
     */
    children: EntityActionExtensionMenuItem[];
}
/**
 * This represents menu information for an entity action.
 */
export declare type EntityActionExtensionMenuEntry = EntityActionExtensionMenuItem | EntityActionExtensionSubmenu;
/**
 * Every component referenced by an entity action extension point must inherit from this.
 */
export declare abstract class _EntityActionExtensionComponent {
    /**
     * This returns is an Observable that defines the menu entry for this component. This is actively bound -
     * the component is able to change it in order to have the menu change (for example, to change busy
     * or enabled states).
     */
    abstract getMenuEntry(entityUrn: string): Observable<EntityActionExtensionMenuEntry>;
    /**
     * This field is a method that is called if a menu item is clicked while enabled. It will be called with the menu
     * item's URN and the URN of the entity that the action is being called for (e.g. the URN of the VM being edited).
     * It must return a single-shot (e.g. of, fromPromise or .first) Observable on completion. This Observable returns
     * a single value, refreshRequested. If this is true, the entity being edited will be immediately refreshed.
     * @param menuItemUrn the URN of the clicked menu item
     * @param entityUrnThe URN of the entity that the action is being called for
     */
    abstract performAction(menuItemUrn: string, entityUrn: string): Observable<{
        refreshRequested: boolean;
    }>;
}
export declare const EntityActionExtensionComponent: typeof _EntityActionExtensionComponent;
```

Since then it works.

**Comment by jmoroski** [comment]
Ahh...build time. Got it. That needs to be addressed.

What other issues did you run into getting @vcd/sdk working with a seed that used a newer Angular version? At the very least, the peer dependencies of the SDK need to be expanded. Is there anything else you encountered? I can try to prioritize a new version for the issues.

**Comment by jhiemer** [comment]
@jmoroski nope besides that, now other issues. Looks good after that change.

Do you guys have an ETA for a new release?

**Comment by jmoroski** [comment]
Should've closed this one out awhile ago. You should now be able to use version 0.11.0 of @vcd/sdk to build in Angular 6+ environment
